### ISSUE TITLE
dart2js: Fail to build with O3

### ISSUE THREAD

**Issue by jimmyshiau** [dart-lang/sdk#36137]
* Dart VM version: 2.2.0 (Tue Feb 26 15:04:32 2019 +0100) on "macos_x64"
* MacOSX
* Chrome

### main.dart
```
import 'dart:async';

void main() {

  _getCache(() {
      return null;
  });
}

Future<T> _getCache<T>(FutureOr<T> then()) {

  return cache.get('', () async {
    return then();
  });
}

abstract class Cache {
  Future<T> get<T>(String id, [FutureOr<T> then()]);
}

Cache cache;
```

* Build success with O2
  `dart2js -O2 -m -oweb/main.dart.js web/main.dart `
* Fail to build with O3
  `dart2js -O3 -m -oweb/main.dart.js web/main.dart `
```
web/main.dart:10:11:
Internal Error: Runtime type information not available for type_variable_local(_getCache.T) in (j:anonymous_closure_local(_getCache_closure)) for j:closure_call(_getCache_closure.call).
Future<T> _getCache<T>(FutureOr<T> then()) {
          ^
The compiler is broken.

When compiling the above element, the compiler crashed. It is not
possible to tell if this is caused by a problem in your program or
not. Regardless, the compiler should not crash.

The Dart team would greatly appreciate if you would take a moment to
report this problem at http://dartbug.com/new.

Please include the following information:

* the name and version of your operating system,

* the Dart SDK build number (2.2.0), and

* the entire message you see here (including the full stack trace
  below as well as the source location above).

The compiler crashed: Assertion failure: Runtime type information not available for type_variable_local(_getCache.T) in (j:anonymous_closure_local(_getCache_closure)) for j:closure_call(_getCache_closure.call).
#0      failedAt (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/diagnostics/invariant.dart:49:3)
#1      LocalsHandler.readLocal (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/ssa/locals_handler.dart:325:11)
#2      TypeBuilder.addTypeVariableReference.readAsParameter (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/ssa/type_builder.dart:143:12)
#3      TypeBuilder.addTypeVariableReference (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/ssa/type_builder.dart:154:31)
#4      TypeBuilder.analyzeTypeArgument (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/ssa/type_builder.dart:224:14)
#5      KernelSsaGraphBuilder.buildGenerator (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/ssa/builder_kernel.dart:1124:30)
#6      KernelSsaGraphBuilder.buildFunctionNode (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/ssa/builder_kernel.dart:1024:7)
#7      KernelSsaGraphBuilder.build.<anonymous closure> (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/ssa/builder_kernel.dart:239:13)
#8      CompilerDiagnosticReporter.withCurrentElement (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/compiler.dart:782:15)
#9      KernelSsaGraphBuilder.build (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/ssa/builder_kernel.dart:211:21)
#10     KernelSsaBuilder.build.<anonymous closure> (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/js_model/js_strategy.dart:204:22)
#11     CompilerTask.measure (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/common/tasks.dart:62:51)
#12     KernelSsaBuilder.build (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/js_model/js_strategy.dart:191:17)
#13     SsaBuilderTask.build (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/ssa/ssa.dart:96:21)
#14     SsaFunctionCompiler.compile (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/ssa/ssa.dart:44:29)
#15     JavaScriptBackend.codegen (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/js_backend/backend.dart:679:26)
#16     KernelCodegenWorkItem.run (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/js_model/js_strategy.dart:174:21)
#17     Compiler.emptyQueue.<anonymous closure>.<anonymous closure>.<anonymous closure>.<anonymous closure>.<anonymous closure> (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/compiler.dart:457:70)
#18     CompilerTask.measureSubtask (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/common/tasks.dart:177:35)
#19     Compiler.emptyQueue.<anonymous closure>.<anonymous closure>.<anonymous closure>.<anonymous closure> (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/compiler.dart:457:32)
#20     CompilerTask.measureSubtask (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/common/tasks.dart:177:35)
#21     Compiler.emptyQueue.<anonymous closure>.<anonymous closure>.<anonymous closure> (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/compiler.dart:455:28)
#22     CompilerDiagnosticReporter.withCurrentElement (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/compiler.dart:782:15)
#23     Compiler.emptyQueue.<anonymous closure>.<anonymous closure> (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/compiler.dart:453:18)
#24     CodegenEnqueuer._forEach (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/js_backend/enqueuer.dart:241:12)
#25     CodegenEnqueuer.forEach (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/js_backend/enqueuer.dart:256:5)
#26     Compiler.emptyQueue.<anonymous closure> (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/compiler.dart:449:16)
#27     CompilerTask.measureSubtask (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/common/tasks.dart:177:35)
#28     Compiler.emptyQueue (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/compiler.dart:448:14)
#29     Compiler.processQueue.<anonymous closure> (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/compiler.dart:474:7)
#30     CompilerTask.measureSubtask (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/common/tasks.dart:177:35)
#31     Compiler.processQueue (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/compiler.dart:467:14)
#32     Compiler.generateJavaScriptCode (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/compiler.dart:376:5)
#33     Compiler.compileFromKernel.<anonymous closure> (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/compiler.dart:420:9)
#34     CompilerTask.measureSubtask (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/common/tasks.dart:177:35)
#35     Compiler.compileFromKernel (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/compiler.dart:394:14)
#36     Compiler.runInternal (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/compiler.dart:281:13)
<asynchronous suspension>
#37     Compiler.run.<anonymous closure>.<anonymous closure> (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/compiler.dart:223:38)
#38     new Future.sync (dart:async/future.dart:224:31)
#39     Compiler.run.<anonymous closure> (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/compiler.dart:223:20)
#40     CompilerTask.measureSubtask (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/common/tasks.dart:177:35)
#41     Compiler.run (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/compiler.dart:220:41)
#42     CompilerImpl.run.<anonymous closure>.<anonymous closure> (file:///b/s/w/ir/cache/builder/sdk/pkg/compiler/lib/src/apiimpl.dart:88:22)
#43     _RootZone.runUnary (dart:async/zone.dart:1379:54)
#44     _FutureListener.handleValue (dart:async/future_impl.dart:126:18)
#45     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:639:45)
#46     Future._propagateToListeners (dart:async/future_impl.dart:668:32)
#47     Future._completeWithValue (dart:async/future_impl.dart:483:5)
#48     Future._asyncComplete.<anonymous closure> (dart:async/future_impl.dart:513:7)
#49     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
#50     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
#51     _runPendingImmediateCallback (dart:isolate/runtime/libisolate_patch.dart:115:13)
#52     _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:172:5)

```

**Comment by jimmyshiau** [comment]
### Workaround 
```
Future<T> _getCache<T>(FutureOr<T> then()) {
  return cache.get('', () {
    return then();
  });
}
```
or
```
Future<T> _getCache<T>(FutureOr<T> then()) {

  return cache.get('', () async {
    //final ok = await check();
    return then() as FutureOr<T>;
  });
}
```

**Comment by sigmundch** [comment]
Thanks @jimmyshiau for reporting. And like before, thank you so much for the small repro.

/cc @johnniwinther

**Comment by jimmyshiau** [comment]
Confirmed fixed with Dart 1.16.1
